File: lua/exercises.lua
function change(amount)
  if math.type(amount) ~= "integer" then
    error("Amount must be an integer")
  end
  if amount < 0 then
    error("Amount cannot be negative")
  end
  local counts, remaining = {}, amount
  for _, denomination in ipairs({25, 10, 5, 1}) do
    counts[denomination] = remaining // denomination
    remaining = remaining % denomination
  end
  return counts
end

-- Write your first then lower case function here
function first_then_lower_case(tbl, predicate)
  for _, value in ipairs(tbl) do
    if predicate(value) then
      return string.lower(value)
    end
  end
-- Return nil if there is no match found
  return nil
end

-- Write your powers generator here
function powers_generator(base, limit)
  return coroutine.create(function()
    local result = 1
    while result <= limit do
      coroutine.yield(result)
      result = result * base
    end
  end)
end

-- Write your say function here
function say(word)
  local words = word or ""
  return function(next_word)
    if next_word then
      -- If next_word is an empty string, add a space
      if next_word == "" then
        words = words .. " "
      else
        -- Remove leading and trailing spaces from next_word
        next_word = next_word:match("^%s*(.-)%s*$")
        -- Only append non-empty words
        if #next_word > 0 then
          words = words .. (words == "" and "" or " ") .. next_word
        end
      end
      return say(words)
    else
      -- Return full sentence when no more words are passed
      return words
    end
  end
end

-- Write your line count function here
function meaningful_line_count(filename)
  local file, err = io.open(filename, "r")
  if not file then
    error("No such file")
  end

  local count = 0
  for line in file:lines() do
    if line:match("%S") and not line:match("^%s*#") then
      count = count + 1
    end
  end

  file:close()
  return count
end

-- Write your Quaternion table here
Quaternion = {}
Quaternion.__index = Quaternion

function Quaternion.new(a, b, c, d)
  local self = setmetatable({}, Quaternion)
  self.a = a
  self.b = b
  self.c = c
  self.d = d
  return self
end

function Quaternion.__add(q1, q2)
  return Quaternion.new(q1.a + q2.a, q1.b + q2.b, q1.c + q2.c, q1.d + q2.d)
end

function Quaternion.__mul(q1, q2)
  return Quaternion.new(
    q1.a * q2.a - q1.b * q2.b - q1.c * q2.c - q1.d * q2.d,  -- Real component
    q1.a * q2.b + q1.b * q2.a + q1.c * q2.d - q1.d * q2.c,  -- i component
    q1.a * q2.c - q1.b * q2.d + q1.c * q2.a + q1.d * q2.b,  -- j component
    q1.a * q2.d + q1.b * q2.c - q1.c * q2.b + q1.d * q2.a   -- k component
  )
end

function Quaternion:coefficients()
  return {self.a, self.b, self.c, self.d}
end

function Quaternion:conjugate()
  return Quaternion.new(self.a, -self.b, -self.c, -self.d)
end

local epsilon = 1e-9
function Quaternion.__eq(q1, q2)
  return math.abs(q1.a - q2.a) < epsilon and math.abs(q1.b - q2.b) < epsilon and
         math.abs(q1.c - q2.c) < epsilon and math.abs(q1.d - q2.d) < epsilon
end

function Quaternion:__tostring()
  local parts = {}

  -- Real component (a)
  if self.a ~= 0 then
    table.insert(parts, tostring(self.a))
  end

  -- i component (b)
  if self.b ~= 0 then
    table.insert(parts, (self.b < 0 and tostring(self.b) or (#parts > 0 and "+" or "") .. tostring(self.b)) .. "i")
  end

  -- j component (c)
  if self.c ~= 0 then
    table.insert(parts, (self.c < 0 and tostring(self.c) or (#parts > 0 and "+" or "") .. tostring(self.c)) .. "j")
  end

  -- k component (d)
  if self.d ~= 0 then
    table.insert(parts, (self.d < 0 and tostring(self.d) or (#parts > 0 and "+" or "") .. tostring(self.d)) .. "k")
  end

  -- If all parts are zero, return "0"
  if #parts == 0 then
    return "0"
  else
    return table.concat(parts)
  end
end

File: swift/exercises.swift
import Foundation

struct NegativeAmountError: Error {}
struct NoSuchFileError: Error {}

func change(_ amount: Int) -> Result<[Int:Int], NegativeAmountError> {
    if amount < 0 {
        return .failure(NegativeAmountError())
    }
    var (counts, remaining) = ([Int:Int](), amount)
    for denomination in [25, 10, 5, 1] {
        (counts[denomination], remaining) = 
            remaining.quotientAndRemainder(dividingBy: denomination)
    }
    return .success(counts)
}

// Write your first then lower case function here

 func firstThenLowerCase(of strings: [String], satisfying predicate: (String) -> Bool) -> String? {
    for string in strings {
        if predicate(string) {
            return string.lowercased() 
        }
    }
    return nil
}

// Write your say function here
class Say {
    private var words: [String]

    init(_ initialWord: String = "") {
        self.words = [initialWord]
    }

    private init(words: [String]) {
        self.words = words
    }

    func and(_ word: String) -> Say {
        var newWords = self.words
        newWords.append(word)
        return Say(words: newWords)
    }

    var phrase: String {
        return self.words.joined(separator: " ")
    }
}

func say(_ initialWord: String = "") -> Say {
    return Say(initialWord)
}


// Write your meaningfulLineCount function here
enum FileError: Error {
    case noSuchFile
}

func meaningfulLineCount(_ filename: String) async -> Result<Int, FileError> {
    do {
        let fileURL = URL(fileURLWithPath: filename)
        let fileContents = try await String(contentsOf: fileURL)
        let lines = fileContents.split(separator: "\n")
        let meaningfulLines = lines.filter { line in
            let trimmedLine = line.trimmingCharacters(in: .whitespaces)
            return !trimmedLine.isEmpty && !trimmedLine.hasPrefix("#")
        }
        return .success(meaningfulLines.count)
    } catch {
        return .failure(.noSuchFile)
    }
}


// Write your Quaternion struct here

struct Quaternion: Equatable {
    let a: Double
    let b: Double
    let c: Double
    let d: Double

    var coefficients: [Double] {
        return [a, b, c, d]
    }

    var conjugate: Quaternion {
        return Quaternion(a: a, b: -b, c: -c, d: -d)
    }

    init(a: Double = 0, b: Double = 0, c: Double = 0, d: Double = 0) {
        self.a = a
        self.b = b
        self.c = c
        self.d = d
    }

    static let ZERO = Quaternion()
    static let I = Quaternion(b: 1)
    static let J = Quaternion(c: 1)
    static let K = Quaternion(d: 1)

    static func +(lhs: Quaternion, rhs: Quaternion) -> Quaternion {
        return Quaternion(a: lhs.a + rhs.a, b: lhs.b + rhs.b, c: lhs.c + rhs.c, d: lhs.d + rhs.d)
    }

    static func *(lhs: Quaternion, rhs: Quaternion) -> Quaternion {
        let a = lhs.a * rhs.a - lhs.b * rhs.b - lhs.c * rhs.c - lhs.d * rhs.d
        let b = lhs.a * rhs.b + lhs.b * rhs.a + lhs.c * rhs.d - lhs.d * rhs.c
        let c = lhs.a * rhs.c - lhs.b * rhs.d + lhs.c * rhs.a + lhs.d * rhs.b
        let d = lhs.a * rhs.d + lhs.b * rhs.c - lhs.c * rhs.b + lhs.d * rhs.a
        return Quaternion(a: a, b: b, c: c, d: d)
    }

    static func ==(lhs: Quaternion, rhs: Quaternion) -> Bool {
        return lhs.a == rhs.a && lhs.b == rhs.b && lhs.c == rhs.c && lhs.d == rhs.d
    }

    var description: String {
        var components: [String] = []
        if a != 0 { components.append("\(a)") }
        if b != 0 { components.append(b < 0 ? "\(b)i" : "+\(b)i") }
        if c != 0 { components.append(c < 0 ? "\(c)j" : "+\(c)j") }
        if d != 0 { components.append(d < 0 ? "\(d)k" : "+\(d)k") }
        if components.isEmpty { return "0" }
        let result = components.joined()
        return result.hasPrefix("+") ? String(result.dropFirst()) : result
    }
}

// Conform to CustomStringConvertible for string representation
extension Quaternion: CustomStringConvertible {}

// Write your Binary Search Tree enum here
